tsfc/finatinterface.py

```diff
@@ -0,0 +1,155 @@
+
+
+__all__ = ("create_element", "supported_elements", "as_fiat_cell")
+
+
+supported_elements = {
+    # These all map directly to FInAT elements
+    "Brezzi-Douglas-Marini": finat.BrezziDouglasMarini,
+    "Brezzi-Douglas-Fortin-Marini": finat.BrezziDouglasFortinMarini,
+    "Discontinuous Lagrange": finat.DiscontinuousLagrange,
+    "Discontinuous Raviart-Thomas": finat.DiscontinuousRaviartThomas,
+    "Lagrange": finat.Lagrange,
+    "Nedelec 1st kind H(curl)": finat.Nedelec,
+    "Nedelec 2nd kind H(curl)": finat.NedelecSecondKind,
+    "Raviart-Thomas": finat.RaviartThomas,
+    "Regge": finat.Regge,
+    # These require special treatment below
+    "DQ": None,
+    "Q": None,
+}
+"""A :class:`.dict` mapping UFL element family names to their
+FIAT-equivalent constructors.  If the value is ``None``, the UFL
+element is supported, but must be handled specially because it doesn't
+have a direct FIAT equivalent."""
+
+
+class FiatElementWrapper(FiatElementBase):
+    def __init__(self, element, degree=None):
+        super(FiatElementWrapper, self).__init__(element)
+        self._degree = degree
+
+    @property
+    def degree(self):
+        return self._degree or super(FiatElementWrapper, self).degree
+
+
+def fiat_compat(element):
+    from tsfc.fiatinterface import create_element
+    return FiatElementWrapper(create_element(element),
+                              degree=spanning_degree(element))
+
+
+@singledispatch
+def convert(element):
+    """Handler for converting UFL elements to FIAT elements.
+
+    :arg element: The UFL element to convert.
+
+    Do not use this function directly, instead call
+    :func:`create_element`."""
+    if element.family() in supported_elements:
+        raise ValueError("Element %s supported, but no handler provided" % element)
+    return fiat_compat(element)
+
+
+# Base finite elements first
+@convert.register(ufl.FiniteElement)
+def convert_finiteelement(element):
+    cell = as_fiat_cell(element.cell())
+    if element.family() not in supported_elements:
+        return fiat_compat(element)
+    lmbda = supported_elements.get(element.family())
+    if lmbda is None:
+        if element.cell().cellname() != "quadrilateral":
+            raise ValueError("%s is supported, but handled incorrectly" %
+                             element.family())
+        # Handle quadrilateral short names like RTCF and RTCE.
+        element = reconstruct_element(element,
+                                      element.family(),
+                                      quad_tpc,
+                                      element.degree())
+        return finat.QuadrilateralElement(create_element(element))
+    return lmbda(cell, element.degree())
+
+
+# VectorElement case
+@convert.register(ufl.VectorElement)
+def convert_vectorelement(element):
+    scalar_element = create_element(element.sub_elements()[0])
+    return finat.TensorFiniteElement(scalar_element, (element.num_sub_elements(),))
```

Comment:
No, because if IIRC `ufl.TensorElement` asserts that the inner element is scalar, while `ufl.VectorElement` gladly accepts an RT element for example.